File: src/lib/permissions.ts
export const SUPER_ADMIN_EMAIL = "somdomato@somdomato.com";

export const ALL_PERMISSIONS = [
  "songs:edit_tags",
  "songs:edit_file",
  "songs:delete",
  "requests:manage",
  "uploads:manage",
  "logs:view",
  "users:manage",
] as const;

export type Permission = (typeof ALL_PERMISSIONS)[number];

export const PERMISSION_LABELS: Record<Permission, string> = {
  "songs:edit_tags": "Editar Tags ID3",
  "songs:edit_file": "Editar Arquivo",
  "songs:delete": "Apagar Música",
  "requests:manage": "Gerenciar Pedidos",
  "uploads:manage": "Gerenciar Envios",
  "logs:view": "Ver Logs",
  "users:manage": "Gerenciar Usuários",
};

export const PERMISSION_GROUPS: Record<string, Permission[]> = {
  Músicas: ["songs:edit_tags", "songs:edit_file", "songs:delete"],
  Pedidos: ["requests:manage"],
  Envios: ["uploads:manage"],
  Logs: ["logs:view"],
  Usuários: ["users:manage"],
};

export const ALL_ROLES = [
  "user",
  "locutor",
  "moderator",
  "admin",
  "super_admin",
] as const;
export type Role = (typeof ALL_ROLES)[number];

export const ADMIN_ROLES: readonly Role[] = [
  "locutor",
  "moderator",
  "admin",
  "super_admin",
];

export const ROLE_LABELS: Record<Role, string> = {
  user: "Ouvinte",
  locutor: "Locutor",
  moderator: "Moderador",
  admin: "Admin",
  super_admin: "Super Admin",
};

export const DEFAULT_ROLE_PERMISSIONS: Record<string, Permission[]> = {
  admin: [
    "songs:edit_tags",
    "songs:edit_file",
    "songs:delete",
    "requests:manage",
    "uploads:manage",
    "logs:view",
    "users:manage",
  ],
  moderator: ["uploads:manage", "requests:manage", "logs:view"],
  locutor: ["requests:manage"],
};

/** Check if a role has any songs-related permission */
export function hasSongsAccess(permissions: string[]): boolean {
  return permissions.some((p) =>
    ["songs:edit_tags", "songs:edit_file", "songs:delete"].includes(p),
  );
}

File: api/rbac/permissions.go
package rbac

// Permission — porta de src/lib/permissions.ts. Os valores atribuídos por
// papel vivem na tabela `role_permissions` (seed em migration 0001); esta
// lista é usada para validar entradas e rotular a UI do admin.
type Permission string

const (
	PermSongsEditTags  Permission = "songs:edit_tags"
	PermSongsEditFile  Permission = "songs:edit_file"
	PermSongsDelete    Permission = "songs:delete"
	PermRequestsManage Permission = "requests:manage"
	PermJinglesManage  Permission = "jingles:manage"
	PermLogsView       Permission = "logs:view"
	PermUsersManage    Permission = "users:manage"
)

var PermissionLabels = map[Permission]string{
	PermSongsEditTags:  "Editar Tags",
	PermSongsEditFile:  "Editar Arquivo",
	PermSongsDelete:    "Apagar Música",
	PermRequestsManage: "Gerenciar Pedidos",
	PermJinglesManage:  "Gerenciar Vinhetas",
	PermLogsView:       "Ver Logs",
	PermUsersManage:    "Gerenciar Usuários",
}

type Role string

const (
	RoleUser       Role = "user"
	RoleLocutor    Role = "locutor"
	RoleModerator  Role = "moderator"
	RoleAdmin      Role = "admin"
	RoleSuperAdmin Role = "super_admin"
)

var AllRoles = []Role{RoleUser, RoleLocutor, RoleModerator, RoleAdmin, RoleSuperAdmin}

var AdminRoles = map[Role]bool{
	RoleLocutor: true, RoleModerator: true, RoleAdmin: true, RoleSuperAdmin: true,
}

var RoleLabels = map[Role]string{
	RoleUser:       "Ouvinte",
	RoleLocutor:    "Locutor",
	RoleModerator:  "Moderador",
	RoleAdmin:      "Admin",
	RoleSuperAdmin: "Super Admin",
}

func IsValidRole(r string) bool {
	for _, v := range AllRoles {
		if string(v) == r {
			return true
		}
	}
	return false
}

func IsAdminRole(r string) bool {
	return AdminRoles[Role(r)]
}
